helix-term/src/commands.rs

```diff
@@ -4240,3 +4300,141 @@ fn suspend(_cx: &mut Context) {
+
+    let breakpoint = helix_dap::SourceBreakpoint {
+        line: text.char_to_line(pos) + 1, // convert from 0-indexing to 1-indexing (TODO: could set debugger to 0-indexing on init)
+        ..Default::default()
+    };
+
+    let path = match doc.path() {
+        Some(path) => path.to_path_buf(),
+        None => {
+            cx.editor
+                .set_error("Can't set breakpoint: document has no path".to_string());
+            return;
+        }
+    };
+
+    // TODO: need to map breakpoints over edits and update them?
+    // we shouldn't really allow editing while debug is running though
+
+    if let Some(debugger) = &mut cx.editor.debugger {
+        let breakpoints = debugger.breakpoints.entry(path.clone()).or_default();
+        if let Some(pos) = breakpoints.iter().position(|b| b.line == breakpoint.line) {
+            breakpoints.remove(pos);
+        } else {
+            breakpoints.push(breakpoint);
+        }
+
+        let breakpoints = breakpoints.clone();
+
+        let request = debugger.set_breakpoints(path, breakpoints);
+        let _ = block_on(request).unwrap();
+    }
+}
+
+fn dap_run(cx: &mut Context) {
+    use helix_lsp::block_on;
+
+    if let Some(debugger) = &mut cx.editor.debugger {
+        if debugger.is_running {
+            cx.editor
+                .set_status("Debuggee is already running".to_owned());
+            return;
+        }
+        let request = debugger.configuration_done();
+        let _ = block_on(request).unwrap();
+        debugger.is_running = true;
+    }
+}
+
+fn dap_continue(cx: &mut Context) {
+    use helix_lsp::block_on;
+
+    if let Some(debugger) = &mut cx.editor.debugger {
+        if debugger.is_running {
+            cx.editor
+                .set_status("Debuggee is already running".to_owned());
+            return;
+        }
+        // assume 0 to continue all threads for now
+        // FIXME: spec conformant behavior here
+        let request = debugger.continue_thread(0);
+        let _ = block_on(request).unwrap();
+        debugger.is_running = true;
+        debugger.stack_pointer = None;
+    }
+}
+
+fn dap_variable_scopes(cx: &mut Context) {
+    use helix_lsp::block_on;
+
+    if let Some(debugger) = &mut cx.editor.debugger {
+        if debugger.is_running {
+            cx.editor
+                .set_status("Cannot access variables while target is running".to_owned());
+            return;
+        }
+        if debugger.stack_pointer.is_none() {
+            cx.editor
+                .set_status("Cannot find current stack pointer to access variables".to_owned());
+            return;
+        }
+
+        let frame_id = debugger.stack_pointer.clone().unwrap().id;
+        let scopes = block_on(debugger.scopes(frame_id)).unwrap();
+
+        let mut status = String::new();
+        for scope in scopes.iter() {
+            status.push_str(&format!("{}: {}; ", scope.variables_reference, scope.name))
+        }
+        cx.editor.set_status(status);
```

Comment:
Current variable presentation is pretty crude, since user has to first obtain scope's variables reference and type it in to get vars. We need two components to make it perfect:
- Menu picker (just shows a list to user to pick from)
- Message window with scroll which shows scrollable text

Reply:
Possibly we can even find those variables in code using LSP and show diagnostics near them for values obtained via debug.

Reply:
Menu picker will also help with attach (show processes)